### PR TITLE
ISSUE-1050 [PublicAsset] PublicAsset/set update JMAP method

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/tmail-backend/issues/1050

### CODE REVIEW

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala [hunk 1]
```diff
@@ -127,4 +131,35 @@ class PublicAssetSetCreatePerformer @Inject()(val publicAssetRepository: PublicA
...
+        case Right((publicAssetId, validatedPath)) => updatePublicAsset(publicAssetId, validatedPath, mailboxSession)
+      }, ReactorUtils.DEFAULT_CONCURRENCY)
+      .collectSeq()
+      .map(PublicAssetUpdateResults)
+
+  private def updatePublicAsset(publicAssetId: PublicAssetId, patch: ValidatedPublicAssetPatchObject, mailboxSession: MailboxSession): SMono[PublicAssetUpdateResult] =
+    publicAssetSetService.checkIdentityIdsExist(patch.identityIds.toSeq, mailboxSession)
+      .flatMap(checkedIdentityIds => SMono(publicAssetRepository.update(mailboxSession.getUser, publicAssetId, checkedIdentityIds.toSet)))
+      .`then`(SMono.just[PublicAssetUpdateResult](PublicAssetUpdateSuccess(publicAssetId)))
+      .onErrorResume(e => SMono.just(PublicAssetUpdateFailure(UnparsedPublicAssetId(publicAssetId.value.toString), e)))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala:153]: Please make sure that the failure is eventually logged.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala [hunk 2]
```diff
@@ -127,4 +131,35 @@ class PublicAssetSetCreatePerformer @Inject()(val publicAssetRepository: PublicA
...
+
+  private def tryGetValidateUpdateRequest(request: PublicAssetSetRequest): SFlux[Either[PublicAssetUpdateFailure, (PublicAssetId, ValidatedPublicAssetPatchObject)]] =
+    SFlux.fromIterable(request.update.getOrElse(Map.empty))
+      .map({
+        case (unparsedPublicAssetId: UnparsedPublicAssetId, patch: PublicAssetPatchObject) =>
+          (for {
+            publicAssetId <- unparsedPublicAssetId.tryAsPublicAssetId
+            validatedPath <- patch.validate()
+          } yield (publicAssetId, validatedPath))
+            .left.map(e => PublicAssetUpdateFailure(unparsedPublicAssetId, e))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala:163]: If I specify an invalid JSON, do I have a server log about it?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetSetRequest.scala [hunk 3]
```diff
@@ -119,4 +128,87 @@ case class PublicAssetCreationResults(created: Seq[PublicAssetCreationResult]) {
...
+case class ValidatedPublicAssetPatchObject(identityIds: Seq[IdentityId])
+
+case class PublicAssetUpdateResponse()
+
+sealed trait PublicAssetUpdateResult
+
+case class PublicAssetUpdateFailure(id: UnparsedPublicAssetId, exception: Throwable) extends PublicAssetUpdateResult {
+  def asSetError: SetError = exception match {
+    case e: PublicAssetException => SetError.invalidArguments(SetErrorDescription(e.getMessage))
+    case e: IllegalArgumentException => SetError.invalidArguments(SetErrorDescription(e.getMessage))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetSetRequest.scala:193]: logs likely go here?

**Review by chibenwa — COMMENTED** [review]
This work looks overall good to me with one reserve, that is more related to the spec part...

The issues I get is that lists updates are non-idenmpotent: to actually express something like `add idc to the list of identities` we actually need to:

 - Fetch the values with PublicAsset/get
 - Set again the list with PublicAsset/set update

And if there is anything wrong with client side caching then the 7 days old client might erase data specified since then. This was a bug happening on 2020 on filter updates.

The classing way to work around this is to model the data as a set:

```
           |          "$publicAssetId": {
           |            "identityIds": {
           |               "id1": true,
           |               "id2": true
           |            }
           |          }
```

This allows to express with json pointers...

... addition:

```
           |          "$publicAssetId": {
           |            "identityIds/id3": true
           |           } 
```

... removal:

```
           |          "$publicAssetId": {
           |            "identityIds/id1": false
           |           } 
```

And of course full reset:

```
           |          "$publicAssetId": {
           |            "identityIds": {
           |               "id2": true,
           |               "id3": true
           |            }
           |          }
```

Using the set datastructure allows expressing those operation without involving client state and is thus more reliable. With DB support, one can also implement a fully concurrency proofed CRDT version of this! (likely overkill, I think having the backend do a get + patch + reset under the hood is plenty enough!)

This is the model used in `Email` keyword property.

What does the team think about adopting this?

If so we would need to adapt the spec (easy), PublicAsset/get (we can do a map serialisation trick to not change too much code), publicAsset/set create (again a simple deserialization trick with a mapRead would do) and of course this PR that would need to be reworked in order to be json-pointer compliant...

Thoughts?

**Comment by vttranlina** [comment]
I agree, now the update method looks like "reset identityId list", it is not easy when we want to add/remove one identityId

**Comment by quantranhong1999** [comment]
> What does the team think about adopting this?

+1

**Comment by Arsnael** [comment]
+1

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala (hunk 1 above)
Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala:153]: I did it in `PublicAssetUpdateFailure.asSetError` method
```
 def asSetError: SetError = exception match {
   ....
    case _ =>
      LOGGER.warn("Unexpected exception when update public asset ", exception)
      SetError.serverFail(SetErrorDescription(exception.getMessage))
  }
  
  ```

**Comment by Arsnael** [comment]
Spec modification => https://github.com/linagora/tmail-backend/pull/1070#pullrequestreview-2080340783

It's ok to refactor the code here to do the full reset, and do partial upgrade in an other PR.

Might want to change /set create first though to fit the new identityIds representation

**Comment by vttranlina** [comment]
I updated code for new spec at commit: https://github.com/linagora/tmail-backend/pull/1063/commits/06125be43128a399ded6da2806bb2d82111e8738
(this pr)

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetSetRequest.scala (hunk 3 above)
Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetSetRequest.scala]: updated log here

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala (hunk 2 above)
Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala:179]: log at PublicAssetUpdateFailure.asSetError

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/json/PublicAssetSerializerTest.scala [hunk 4]
```diff
@@ -42,6 +43,30 @@ class PublicAssetSerializerTest {
...
   }
 
+  @Test
+  def deserializePublicAssetSetUpdateRequestShouldSucceed(): Unit = {
+    val jsInput: JsValue = Json.parse(
+      """{
+        |  "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+        |  "update": {
+        |    "4f29": {
+        |      "identityIds": ["12", "34"]
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/json/PublicAssetSerializerTest.scala:53]: Oups?

```suggestion
        |      "identityIds": {"12":true, "34":true}
```

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**